### DISCUSSION TITLE
Eleventy Edge - Nothing being displayed between tags on Netlify but works when running netlify dev locally

### DISCUSSION THREAD

**Discussion by mark-reason** [11ty/eleventy#2409] [Uncategorized]
Hi,

I'm using the generated eleventy-edge function with an extra cookie added and the filter changed:
``` javascript
import { EleventyEdge } from "eleventy:edge";
import precompiledAppData from "./_generated/eleventy-edge-app-data.js";

export default async (request, context) => {
  try {
    console.log("Eleventy Edge function running");
    let edge = new EleventyEdge("edge", {
      request,
      context,
      precompiled: precompiledAppData,

      // default is [], add more keys to opt-in e.g. ["appearance", "username"]
      cookies: ["bv-cookie-policy"],
    });

    edge.config((eleventyConfig) => {
      // Add some custom Edge-specific configuration
      // e.g. Fancier json output
      eleventyConfig.addFilter("json", (string) => JSON.parse(string));
    });

    return await edge.handleResponse();
  } catch (e) {
    console.log("ERROR", { e });
    return context.next(e);
  }
};
```

This is being used to decide whether a cookie banner is shown or not:
``` html
{% edge %}
{% set cookiePolicy = eleventy.edge.cookies['bv-cookie-policy'] %}

{% if not cookiePolicy %}
	<div id="cookie-banner">
		<div class="container">
			<h2>Let us know you agree to cookies</h2>
			<p>We use some essential cookies to make this website work. We'd like to set additional cookies to give you the best online experience.</p>

			<div><button type="button" data-button="accept-cookies">Accept all cookies</button></div>
			<div><a href="/manage-cookies" class="btn">Manage cookies</a></div>
		</div>
	</div>
{% endif %}
{% endedge %}
```

and this is my netlify.toml (there is another Netlify function in there that handles site search, this is working fine)
```
[build]
  command = "npm run prod"
  publish = "dist"

[[headers]]
  for = "/*"
  [headers.values]
    X-Frame-Options = "DENY"
    X-XSS-Protection = "1; mode=block"
    X-Content-Type-Options = "nosniff"
    Referrer-Policy = "same-origin"
    Permissions-Policy = "microphone=(self), camera=(self)"
    Content-Security-Policy = "upgrade-insecure-requests;"

[[edge_functions]]
  function = "site-search"
  path = "/search"

[[edge_functions]]
  function = "site-search"
  path = "/search/"

[[edge_functions]]
  function = "eleventy-edge"
  path = "/*"
```

If I run netlify dev locally and the cookie I'm looking for doesn't exist then the cookie banner content will show. However when I deploy this to an actual Netlify server nothing shows at all. The function is running as I can see the console log in the Netlify edge function log.

Package versions:
"@11ty/eleventy": "^v2.0.0-canary.7"
"@netlify/functions": "^1.0.0"

Am I doing anything wrong here?

Cheers
Mark

**Comment by zachleat** [comment]
So it works locally but not on Netlify servers, hmm. Which type of Netlify account are you using? https://www.netlify.com/pricing/

Can you share the code that writes the `bv-cookie-policy` cookie? And/or the full value of the cookie from your devtools on both localhost and Netlify server?

Also probably unrelated but just to eliminate variables here can you update to the latest canary? I think we’re up to `11` as of posting this.

(note for me, this was also posted to the netlify support forums https://answers.netlify.com/t/eleventy-edge-nothing-being-displayed-between-tags-on-netlify-but-works-when-running-netlify-dev-locally/60315)

**Reply by mark-reason** [reply]
Thanks for the response. I am using a 'Pro' paid account. An example of the value of the cookie is:
```JSON
{"essential":"true","usage":"true","marketing":"true","settings":"true"}
```

Code that sets the cookie...

cookies.js (utility functions)
```JavaScript
/**
 * Set a cookie
 *
 * @param {String} name - name of cookie
 * @param {String} value - value of cookie
 * @param {Number} days - expiry
 */
export const setCookie = (name, value, days) => {
  var expires = "";
  if (days) {
    var date = new Date();
    date.setTime(date.getTime() + days * 24 * 60 * 60 * 1000);
    expires = "; expires=" + date.toUTCString();
  }
  document.cookie = name + "=" + (value || "") + expires + "; path=/";
};

/**
 * Get a cookie value
 *
 * @param {String} name - name of cookie
 */
export const getCookie = (name) => {
  var nameEQ = name + "=";
  var ca = document.cookie.split(";");
  for (var i = 0; i < ca.length; i++) {
    var c = ca[i];
    while (c.charAt(0) === " ") c = c.substring(1, c.length);
    if (c.indexOf(nameEQ) === 0) return c.substring(nameEQ.length, c.length);
  }
  return null;
};

/**
 * Delete a cookie
 *
 * @param {String} name - name of cookie
 */
export const deleteCookie = (name) => {
  document.cookie = name + "=; Path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;";
};
```

cookie-settings.js (sets the cookie policy)
```Javascript
import { setCookie, getCookie } from "../utils/cookies";

const cookieBanner = document.getElementById("cookie-banner");
const form = document.getElementById("cookie-policy-form");
const acceptCookiesButtons = document.querySelectorAll(
  "[data-button=accept-cookies]"
);
const cookieName = "bv-cookie-policy";
const defaultPolicy = {
  essential: "true",
  usage: "false",
  marketing: "false",
  settings: "false",
};
let userPolicy = getCookie(cookieName)
  ? JSON.parse(getCookie(cookieName))
  : defaultPolicy;

/**
 * Change one option in the cookie policy
 *
 * @param {String} name - name of option
 * @param {String} value
 */
const changeCookieOption = (name, value = "true") => {
  userPolicy[name] = value;
};

/**
 * Get user value from cookie policy
 *
 * @param {String} key
 */
export const cookiePolicyValue = (key) => {
  return userPolicy[key];
};

/**
 * Update all cookie policy options to true
 */
const acceptAll = () => {
  changeCookieOption("usage");
  changeCookieOption("marketing");
  changeCookieOption("settings");
  location.reload();
};

/**
 * Handle accept all cookies buttons
 */
if (acceptCookiesButtons?.length > 0) {
  Array.prototype.forEach.call(acceptCookiesButtons, function (btn) {
    btn.addEventListener("click", (e) => {
      e.preventDefault();
      acceptAll();
      setCookie(cookieName, JSON.stringify(userPolicy), 365);
      if (cookieBanner) cookieBanner.hidden = true;
    });
  });
}

/**
 * Handle cookie setting form submit
 */
if (form) {
  form.addEventListener("submit", (e) => {
    e.preventDefault();
    for (const item in userPolicy) {
      const value = document.querySelector(`input[name=${item}]:checked`).value;
      changeCookieOption(item, value);
    }
    setCookie(cookieName, JSON.stringify(userPolicy), 365);
    location.reload();
  });
}
```

and finally I get the same issue running the latest version
```
[11ty] Copied 17 files / Wrote 187 files in 14.04 seconds (75.1ms each, v2.0.0-canary.11)
```

**Reply by zachleat** [reply]
Thanks! Can you also share a devtools screenshot of the cookie values in your browser? My first guess is that the cookie isn’t accessible because of extra cookie attributes?

Firefox
![image](https://user-images.githubusercontent.com/39355/170287564-3e5fd22e-ef2f-46d2-a2fc-391dd36429d9.png)

Chrome:
![image](https://user-images.githubusercontent.com/39355/170287736-d26d0c47-7453-463e-868b-ee858ac54725.png)

**Reply by mark-reason** [reply]
![Screenshot 2022-05-25 at 15 37 20](https://user-images.githubusercontent.com/5191335/170288583-9da64bfe-b98f-47b2-9045-3fb1276cf588.png)

**Reply by zachleat** [reply]
Just re-reading here—I don’t think I took us down the right debugging road here sorry! I got tied up in cookie things but really this is about Edge processing. Does the build output (to `_site`) show the Edge snippets? Does anything show up in your Edge functions logs in the Netlify app?

**Reply by mark-reason** [reply]
Yep I've added a console.log to the bottom of the edge function and this displays in the logs.

There doesn't seem to be any edge output when I view the source of the page when viewing on the Netlify instance

**Reply by zachleat** [reply]
Ah, so if you see the tags in your `_site` folder and not live on the production site that does confirm that it is running the Edge code.

This shouldn’t fork on local versus production but just for completeness I’ll mention https://twitter.com/eleven_ty/status/1525221418262200320

Does something like this render on Netlify?

```
{% edge %}
test
{% endedge %}
```

**Reply by mark-reason** [reply]
Yes so I can see the tags in my output folder locally (<!--ELEVENTYEDGE_edge).

Have tried the test but that doesn't show on production either
![Screenshot 2022-05-27 at 09 13 49](https://user-images.githubusercontent.com/5191335/170659542-e9057de5-07fc-4dd7-8cc5-04d95f95e1c3.png)

Don't know if it's worth mentioning but I has a similar issue with an on demand builder running Eleventy which I fixed by adding this to my netlify.toml

```
[functions]
  included_files = ['src/**/*', 'package.json', 'utils/**/*', '.eleventy.js']
```

That hasn't fixed the edge issue though.

**Reply by zachleat** [reply]
Unfortunately I’m going to need more access to debug the issue to see what’s going on. Is the source code available?

**Reply by mark-reason** [reply]
It's not public unfortunately but I can send privately?

**Reply by zachleat** [reply]
yeah feel free to invite my user to the repo `@zachleat`

**Reply by mark-reason** [reply]
I have sent you an invite

**Reply by mark-reason** [reply]
Okay @zachleat, I think I found the culprit... 

In my .eleventy.js
```
// Only minify HTML if we are in production because it slows builds _right_ down
  if (process.env.ELEVENTY_ENV === "production") {
    config.addTransform(
      "htmlmin",
      require("./utils/transforms/html-min-transform.js")
    );
  }
```

I'm presuming the minification is stripping out the comments that are used as placeholders for the Eleventy Edge function.

**Reply by zachleat** [reply]
Yeah, you’re absolutely right. Some additional discussion here https://github.com/11ty/eleventy/issues/2433

Can you add your HTML minifier to #2433? Thank you!

**Reply by mark-reason** [reply]
I was using html-minifier also so I won't add anything :)
